src/riak_kv_index_hashtree.erl

```
@@ -675,52 +711,98 @@ maybe_build(State) ->
+    case {Locked, Type} of
+        {true, build} ->
+            lager:debug("Starting build: ~p", [Index]),
+            fold_keys(Index, Self),
+            lager:debug("Finished build (a): ~p", [Index]), 
+            gen_server:cast(Self, build_finished);
+        {true, rehash} ->
+            lager:debug("Starting rehash: ~p", [Index]),
+            [hashtree:rehash_tree(T) || {_,T} <- Trees],
+            lager:debug("Finished rehash (a): ~p", [Index]),
+            gen_server:cast(Self, build_finished);
+        _ ->
             gen_server:cast(Self, build_failed)
     end.
 
+-spec maybe_rebuild(state()) -> state().
+maybe_rebuild(State=#state{lock=undefined, built=true, expired=true, index=Index}) ->
+    Self = self(),
+    Pid = spawn_link(fun() ->
+                             receive
+                                 {lock, Locked, State2} ->
+                                     build_or_rehash(Self, Locked, build, State2);
+                                 stop ->
+                                     ok
+                             end
+                     end),
+    Locked = get_all_locks(build, Index, Pid),
+    case Locked of
+        true ->
+            State2 = clear_tree(State),
+            Pid ! {lock, Locked, State2},
+            State2#state{built=Pid};
+        _ ->
+            Pid ! stop,
+            State
+    end;
+maybe_rebuild(State) ->
+    State.
+
 close_trees(State=#state{trees=Trees}) ->
-    Trees2 = [{IdxN, hashtree:close(Tree)} || {IdxN, Tree} <- Trees],
-    State#state{trees=Trees2}.
+    Trees2 = [begin
+                  NewTree = try
+                                hashtree:flush_buffer(Tree)
+                            catch _:_ ->
+                                    Tree
+                            end,
+                  {IdxN, NewTree}
+              end || {IdxN, Tree} <- Trees],
+    Trees3 = [{IdxN, hashtree:close(Tree)} || {IdxN, Tree} <- Trees2],
```

I can see that the current hashtree:close() operations is pretty crash safe. But code does change. Would it make sense to wrap the try catch around the flush and close? I would think whatever makes flush fail, will make close fail, even though right now it would probably just fail silently.